streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -258,23 +256,24 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
 
         for (final Map.Entry<Task, Map<TopicPartition, Long>> taskAndCheckpoint : checkpointPerTask.entrySet()) {
             final Task task = taskAndCheckpoint.getKey();
+            final Map<TopicPartition, Long> checkpoint = taskAndCheckpoint.getValue();
+
             try {
-                task.closeClean(checkpointPerTask.get(task));
+                completeTaskCloseClean(task, checkpoint);
+                cleanUpTaskProducer(task, taskCloseExceptions);
+                tasks.remove(task.id());
             } catch (final RuntimeException e) {
                 final String uncleanMessage = String.format("Failed to close task %s cleanly. Attempting to close remaining tasks before re-throwing:", task.id());
                 log.error(uncleanMessage, e);
                 taskCloseExceptions.put(task.id(), e);
                 // We've already recorded the exception (which is the point of clean).
                 // Now, we should go ahead and complete the close because a half-closed task is no good to anyone.
-                task.closeDirty();
-            } finally {
-                cleanUpTaskProducer(task, taskCloseExceptions);
-                tasks.remove(task.id());
+                dirtyTasks.add(task);
             }
         }
 
         for (final Task task : dirtyTasks) {
-            task.closeDirty();
+            closeTaskDirty(task);
```

Comment:
This was the main bug, we try to batch commit active tasks above and deal with failure by adding to this `dirtyTasks` set. But we didn't actually do all of the required cleanup for a removed task in this loop here, specifically we did not ever call `cleanupTask` on tasks that were added to `dirtyTasks` from the `additionalTasksForCommitting` set on line 250

Reply:
I think we also didn't always call `prepareCloseDirty` before calling `closeDirty`. That was not the source of this particular bug, but I tried to consolidate all the cleanup tasks a bit to make sure they end up being called at all times, in the correct order

Reply:
LGTM.